sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,766 @@
+  // enable_if T is a scalar type
+  T get_on_device() {
+    const char *SymbolicID = __builtin_unique_stable_name(detail::specialization_id_name_generator<S>);
+    return __sycl_getScalar2020SpecConstantValue<T>(SymbolicID, &S, Ptr);
+  }
+
+  template<auto &S, typename T = std::remove_reference_t<decltype(S)>::type>
+  // enable_if T is a composite type
+  T get_on_device() {
+    const char *SymbolicID = __builtin_unique_stable_name(detail::specialization_id_name_generator<S>);
+    return __sycl_getComposite2020SpecConstantValue<T>(SymbolicID, &S, Ptr);
+  }
+#endif // __SYCL_DEVICE_ONLY__
+
+  byte *Ptr = nullptr;
+};
+
+} // namespace sycl
+```
+
+Here [`__builtin_unique_stable_name`][builtin-unique-stable-name]
+is a compiler built-in used to translate types to unique strings, which are
+used as symbolic IDs of specialization constants.
+
+[builtin-unique-stable-name]: https://github.com/intel/llvm/blob/sycl/clang/docs/LanguageExtensions.rst#__builtin_unique_stable_name
+
+`__sycl_getScalar2020SpecConstantValue<T>` and
+`__sycl_getComposite2020SpecConstant<T>` are functions with special names - they
+are declared in the headers but never defined. Calls to them are recognized by
+a special LLVM pass later and this is aforementioned special markup required for
+the compiler.
+Those functions accept three parameters:
+1. Symbolic ID of a specialization constant. Even though at SPIR-V level
+   specialization constants are identified by numeric IDs, we can't use them
+   here, because:
+   - Those IDs can't be generated by runtime, because they need to be encoded
+     into resulting SPIR-V device image
+   - Those IDs can't be generated by front-end compiler, because it only sees a
+     single translation unit at a time and therefore it can't assign unique IDs
+     to specialization constants from different translation units.
+
+   Therefore, the decision was made to use symbolic IDs as interface between the
+   compiler and runtime to connect SYCL identifiers of specialization constants
+   with SPIR-V identifiers of specialization constants.
+
+2. Default value of the specialization constant.
+   It is expected that at LLVM IR level the argument will contain a pointer to
+   a global variable with the initializer, which should be used as the default
+   value of the specialization constants.
+
+3. Pointer to a buffer, which will be used if native specialization constants
+   are not available. This pointer is described later in the section
+   corresponding to emulation of specialization constants.
+
+Compilation and subsequent linkage of the device code results in a number of
+`__sycl_getScalar2020SpecConstantValue` and
+`__sycl_getComposite2020SpecConstantValue` calls. Before generating a device
+binary, each linked device code LLVM IR module undergoes processing by
+`sycl-post-link` tool which can run LLVM IR passes before passing the module
+onto the SPIR-V translator.
+
```

Comment:
What will happen if there is translation from LLVM IR to SPIRV prior to sycl-post-link-tool?  This feature/functionality must work or allow for LLVM IR to SPIRV translation prior to link, potential translation back to LLVM IR for linking, and
then another translation back to SPIRV after sycl-post-link.  Even though this is not the default translation method, allowing SPIRV as a binary format in our initial fat objects should be a design requirement.

Reply:
Theoretically this should work. But this assumes LLVMIR->SPIRV->LLVMIR translation does not loose any significant info (like linkage attributes of globals, `linkonce_odr`), this will likely require fixes in the translator here and there, or even some SPIRV extensions.
Spec constants design itself does not depend directly on fat object format, though, it is the post-link IR transformation model which does. I believe it still needs LLVMIR as the IR format post-link transformations work on, as SPIRV is not really suitable. So I think the spec constants design can assume it has LLVMIR at its disposal.

Reply:
I am OK with LLVM IR being the IR that post-link operates on.   I was particularly concerned about what happens to kernel APIs that resolve/refer to specialization constants if LLVM IR gets converted to SPIRV sometime before the post-link tool.  We need to make sure that LLVM IR -> SPIRV -> LLVMIR round trip preserves anything need for specialization constants to work.

Reply:
It seems the only thing we need is support for `linkonce_odr` in SPIRV, other things are preserved during the round trip.